pkg/reconciler/reconciler.go

```diff
@@ -660,28 +667,27 @@ const (
 )
 
 func (r *Reconciler) handleDeletion(ctx context.Context, actionClient helmclient.ActionInterface, obj *unstructured.Unstructured, log logr.Logger) error {
-	if !controllerutil.ContainsFinalizer(obj, uninstallFinalizer) {
-		log.Info("Resource is terminated, skipping reconciliation")
-		return nil
-	}
-
-	// Use defer in a closure so that it executes before we wait for
-	// the deletion of the CR. This might seem unnecessary since we're
-	// applying changes to the CR after is has a deletion timestamp.
-	// However, if uninstall fails, the finalizer will not be removed
-	// and we need to be able to update the conditions on the CR to
-	// indicate that the uninstall failed.
-	if err := func() (err error) {
-		uninstallUpdater := updater.New(r.client)
-		defer func() {
-			applyErr := uninstallUpdater.Apply(ctx, obj)
-			if err == nil {
-				err = applyErr
-			}
-		}()
-		return r.doUninstall(actionClient, &uninstallUpdater, obj, log)
-	}(); err != nil {
-		return err
+	if controllerutil.ContainsFinalizer(obj, uninstallFinalizer) {
+		// Use defer in a closure so that it executes before we wait for
+		// the deletion of the CR. This might seem unnecessary since we're
+		// applying changes to the CR after is has a deletion timestamp.
+		// However, if uninstall fails, the finalizer will not be removed
+		// and we need to be able to update the conditions on the CR to
+		// indicate that the uninstall failed.
+		if err := func() (err error) {
+			uninstallUpdater := updater.New(r.client)
+			defer func() {
+				applyErr := uninstallUpdater.Apply(ctx, obj)
+				if err == nil {
+					err = applyErr
+				}
+			}()
+			return r.doUninstall(actionClient, &uninstallUpdater, obj, log)
+		}(); err != nil {
+			return err
+		}
+	} else {
+		log.Info("Resource is already terminated, skipping deletion")
```

Comment:
Is this change necessary? I *personally* prefer the "early return" on the original rather than wrapping all the uninstall logic in an if statement. From what I can tell, the logic is still exactly the same as the original so IMO we shouldn't worry about changing it.

Reply:
It's not exactly the same, as the old logic wouldn't wait for the resource to be deleted before returning. I thought that waiting for the deletion in both cases would be better, what do you think?

Reply:
Ah okay - thanks for the clarification, guess I should have expanded to see the entire function :facepalm: